### PR TITLE
Docs for builtin pickers

### PR DESCRIPTION
**Pull request by smithbm2316**
Just adding this PR so that as I continue to work on this in the coming days, people can see this and don't start working on docs for this file. Most of what I added in this first commit will probably change. I will also be adding to the README once I have finalized docgen docs for `telescope.txt` first :smiley:

### CODE REVIEW

**Comment by smithbm2316** [comment]
Pinging @tami5 so you don't work on this in the documentation you were planning to work on

**Comment by tami5** [comment]
😍😍😍 noted 📝

**Comment by smithbm2316** [comment]
@tami5 the first draft for this PR is done. Would love to hear any feedback you have on the actual docs I wrote, how I structured them, or anything else you feel should be re-worded, rewritten, or tweaked. 

It's definitely not ready to merge yet, as the docgen does not seem to be working properly for the files that I added docs to, even after updating the docgen script. I will have to debug this further, maybe TJ or Conni might have an idea what might be going wrong and I can ping them if I don't have any success with figuring out the issue myself. I'll look into that aspect in the coming days.

File: README.md
Comment (tami5) [inline README.md:391]: ctags need to manually generated, I think this is something important to mention in order to manage expectations. 

_Side note: God I always hated this. I'm sure whoever decided to neglect the generation of tags to the user, actually doesn't use tags 😒_

File: README.md
Comment (tami5) [inline README.md:403]: 😯 This is new to me

File: README.md
Comment (tami5) [inline README.md:422]: Lists .... sounds better then searches.

File: lua/telescope/builtin/files.lua
Comment (tami5) [inline lua/telescope/builtin/files.lua:4]: File-related pickers

File: lua/telescope/builtin/files.lua
Comment (tami5) [inline lua/telescope/builtin/files.lua:39]: @Conni2461 please correct me if I'm wrong, the syntax for create fields with docgen is

```
---@field search_dirs table: desc
```

Also @Conni2461 does this auto point to opts?

File: lua/telescope/builtin/git.lua
Comment (tami5) [inline lua/telescope/builtin/git.lua:4]: Git-related pickers

File: lua/telescope/builtin/git.lua
Comment (tami5) [inline lua/telescope/builtin/git.lua:19]: Well it ignore not tracked files as well?

File: lua/telescope/builtin/git.lua
Comment (tami5) [inline lua/telescope/builtin/git.lua:96]: 😍😍

File: lua/telescope/builtin/init.lua
Comment (tami5) [inline lua/telescope/builtin/init.lua:4]: Maybe some rewording is needed here.

e.g. 

> Builtins are a collection of community maintained pickers to support common workflows. It can be used as reference when building prs, extension, or user pickers.

File: lua/telescope/builtin/internal.lua
Comment (tami5) [inline lua/telescope/builtin/internal.lua:27]: Lists builtin pickers ..... 😷

File: lua/telescope/builtin/internal.lua
Comment (tami5) [inline lua/telescope/builtin/internal.lua:60]: @tjdevries where everything began 🥰. I suggest moving this to an examples or scratch folder and save space 😬.

File: lua/telescope/builtin/lsp.lua
Comment (tami5) [inline lua/telescope/builtin/lsp.lua:4]: Neovim Lsp related pickers

**Review by tami5 — COMMENTED** [review]
Amazing work @smithbm2316, I truly appreciate it.

File: README.md
Comment (smithbm2316) [reply README.md:372]: This is just for `git_files` right? Or also for `find_files`?

File: README.md
Comment (smithbm2316) [reply README.md:375]: Will do, this is what the previous entry in the README was so I was gonna leave it since someone added it. I think removing it probably better though as it doesn't really describe much to anyone who hasn't used Emacs before.

File: README.md
Comment (smithbm2316) [reply README.md:391]: `Lists tags in current directory with preview (ctags -R), so long as you manually generated them yourself already`

Do you think that wording is better? ^

File: README.md
Comment (tami5) [reply README.md:372]: > This is just for `git_files` right? Or also for `find_files`? 

I'm referring to git_files. 

find_files I believe also respect .gitignore and ignores files/dirs listed there

File: README.md
Comment (smithbm2316) [reply README.md:403]: Honestly the best part of writing these docs, outside of providing better instructions for the community :smile:, is that I learned about several features and pickers I didn't even know existed! Big win in my book :sunglasses:

File: README.md
Comment (tami5) [reply README.md:391]: > `Lists tags in current directory with preview (ctags -R), so long as you have already manually generated them before`
> 
> 
> 
> Do you think that wording is better? ^

Perhaps:

Lists tags in current directory with tag location file preview (users are required to run `ctags -R` to generate tags or update when introducing new changes)

File: lua/telescope/builtin/files.lua
Comment (smithbm2316) [reply lua/telescope/builtin/files.lua:39]: Ooooh....I may not have read those docgen docs clearly enough :laughing: I'll look at this later tomorrow lol

File: lua/telescope/builtin/git.lua
Comment (smithbm2316) [reply lua/telescope/builtin/git.lua:19]: I think it does ignore untracked files as well, I'll have to find out. I haven't used this picker before myself.

File: lua/telescope/builtin/init.lua
Comment (smithbm2316) [reply lua/telescope/builtin/init.lua:4]: > Builtins are a collection of community maintained pickers to support common workflows. It can be used as reference when writing PRs, Telescope extensions, your own custom pickers, or just as a discovery tool for all of the amazing pickers already shipped with Telescope!

Thoughts on this?

File: lua/telescope/builtin/internal.lua
Comment (smithbm2316) [reply lua/telescope/builtin/internal.lua:60]: The planets picker is my most used picker how dare you!!!! You'll have to pry it out of my cold, dead hands :stuck_out_tongue_closed_eyes:

File: README.md
Comment (smithbm2316) [reply README.md:391]: Sounds good to me!

File: README.md
Comment (smithbm2316) [reply README.md:422]: I'll rewrite all of those and have you look at the README with my next commit again with using "Lists" instead to reword those :)

File: README.md
Comment (smithbm2316) [reply README.md:372]: I'll look into this further and update it accordingly. Will leave this comment open until I have commited something that adds more info for this

File: lua/telescope/builtin/init.lua
Comment (tami5) [reply lua/telescope/builtin/init.lua:4]: Oh that's sound so much better. `Telescope Builtins is a collection of`..

File: lua/telescope/builtin/internal.lua
Comment (tami5) [reply lua/telescope/builtin/internal.lua:60]: 🤣🤣🤣🤣. I doubt that @tjdevries would see this anytime soon. So yah let's move it,

File: lua/telescope/builtin/init.lua
Comment (smithbm2316) [reply lua/telescope/builtin/init.lua:4]: Done!

File: lua/telescope/builtin/internal.lua
Comment (smithbm2316) [reply lua/telescope/builtin/internal.lua:60]: Don't worry, I'll go pester him on stream later this week until he looks only at this specific comment, and ignores everything else about this PR :sunglasses: :rofl:

File: lua/telescope/builtin/git.lua
Comment (smithbm2316) [reply lua/telescope/builtin/git.lua:96]: I'm proud of this one, glad you like it :heart:

File: README.md
Comment (smithbm2316) [reply README.md]: Most recent commit should have updates for this particular note, updated the README with these, and all the LSP README updates should also now be reflected in `builtin/lsp.lua` function headers too

**Comment by Conni2461** [comment]
FYI (and thats what i dreamed about) i wanted to only have documentation in `telescope/builtin/init.lua` (so basically this acts as a header file) and keep the implementations (`telescope/builtin/files.lua`) clean. I see why you want to split it up in the different files (max query result issue) but i am not sure about it

**Comment by smithbm2316** [comment]
> FYI (and thats what i dreamed about) i wanted to only have documentation in `telescope/builtin/init.lua` (so basically this acts as a header file) and keep the implementations (`telescope/builtin/files.lua`) clean. I see why you want to split it up in the different files (max query result issue) but i am not sure about it

For sure! I've never written any C so the idea of "header files" are a bit foreign to me. I wrote it in the actual files because I figured it'd be nice to have the reference right where the function itself was, but if you would prefer for me to keep it in `builtin/init.lua`, I can move everything to there. It was definitely more of a challenge to make sure I covered every picker with docs having to scroll through every file searching for the next function, rather than just managing it in the `init.lua` fil. I'll go ahead and start working on moving the docs back to that file.

**Comment by tjdevries** [comment]
Drive by comment: do not make the docs different just because of max query issue. Upstream knows about it and will fix at some point. We should, for now, just document in our ideal way and then later will get the upstream changes.

I will review more later.

File: README.md
Comment (smithbm2316) [reply README.md:372]: This does indeed _allow you_ to ignore untracked files, but it doesn't look like it does by default. I actually noted this in the function header for `git.files`, but did not here. I can add an extra note to the readme about that.

**Comment by smithbm2316** [comment]
@tami5 I've made all of the changes you've suggested/requested from your review. The only two things left for me to do for this PR is to update the docs to use the proper `@field` and `@param` values instead of the current iteration, and then we need to decide if we want me to move the docs all to `builtin/init.lua` or keep them in the separate file.

**Comment by Conni2461** [comment]
It just can't find the latest nightly image (probably because its untagged). We had this a couple of times in the past, but most of the time the next day it was fixed :laughing: 

My reasoning for why we want to have the documentation in `builtin/init.lua` was to make the actual implementation files `internal.lua`, etc not that cluttered (easier to navigate). Previously we had one file `builtin.lua` and i split it up because developing new builtins was kinda annoying in that +1000 lines file. Now `internal.lua` is almost 1000 lines long again and adding documentation on top of that, will make developing new features not easier.

Also this documentation is not helpful when developing stuff so it shouldn't be part of the implementation but just the definition (init.lua) if that makes sense.

**Comment by smithbm2316** [comment]
That makes perfect sense @Conni2461, I think that does seem like a better option. I'm more than happy to move it to the main `builtin/init.lua` file, I'll try to get that done sometime in the next day or two so we can get this merged and off of the todo list!

**Comment by smithbm2316** [comment]
Alright, I've finished moving the docs to `builtin/init.lua` now @Conni2461 and using the `@param` and `@field` tags where appropriate @tami5. Definitely still having a major issue with the max-query bug here, as most of the docs are not being generated, but the diff is much cleaner now and this should be ready for another review, at least to make sure wording is good and everything changed in the files that don't include `telescope.txt` are ready to go for when the max-query bug is solved upstream.

Does anyone know what the `opts.follow` and `opts.show_line` options do? I wasn't able to figure those ones out, but I think I've mostly covered the important custom ones for all the pickers and added all the custom keymaps that are picker-specific too.

File: lua/telescope/builtin/init.lua
Comment (tami5) [inline lua/telescope/builtin/init.lua:30]: CLEAN 😍 😍

**Comment by tami5** [comment]
> Alright, I've finished moving the docs to `builtin/init.lua` now @Conni2461 and using the `@param` and `@field` tags where appropriate @tami5. Definitely still having a major issue with the max-query bug here, as most of the docs are not being generated, but the diff is much cleaner now and this should be ready for another review, at least to make sure wording is good and everything changed in the files that don't include `telescope.txt` are ready to go for when the max-query bug is solved upstream.
> 
> Does anyone know what the `opts.follow` and `opts.show_line` options do? I wasn't able to figure those ones out, but I think I've mostly covered the important custom ones for all the pickers and added all the custom keymaps that are picker-specific too.

- follow: for following symlinks I believe.
- show_line: for either showing or heading matching lines when doing fuzzy greb or treesitter. 

Not 100% sure, someone correct me

**Review by tami5 — APPROVED** [review]
LGTM

File: lua/telescope/builtin/init.lua
Comment (elianiva) [inline lua/telescope/builtin/init.lua:251]: `..under the cursor which can be triggered by pressing...` is better IMO

File: lua/telescope/builtin/init.lua
Comment (elianiva) [inline lua/telescope/builtin/init.lua:65]: we should also probably tell if it's possible to do `path/to/file.txt` and it will create nested directories. Some people don't know if this is possible, it's useful for Java where you'd have `a/long/nested/dir/for/whatever/Reason.java`

File: lua/telescope/builtin/init.lua
Comment (elianiva) [inline lua/telescope/builtin/init.lua:265]: I think `asynchronously` sounds a bit ambiguous here. It *is* async but it doesn't tell you if you can 'refresh' the result on-the-fly. `dynamically` or `live` is a better word IMO because it's similar to `live_grep`.

**Review by elianiva — COMMENTED** [review]
left some comments, great work :)

**Comment by Conni2461** [comment]
I am doing telescope work this weekend, so i will look at this. Tj and i talked about a workaround for that bug. I will probably also implement that this weekend. Give me one more day :)

**Comment by smithbm2316** [comment]
Finished the last comments that everyone had. Do y'all normally squash and merge normally with PRs here? I also was not sure if we want to merge this yet due to the treesitter bug atm. Happy to hold off on merging until that's sorted if we want.

**Comment by smithbm2316** [comment]
> We do squash and merge. Its in this dropdown so you don't have to do it normally.

Perfect, that's what I assumed. 

> But please wait one more day, i know i am already late but i had other things :rofl: 

I will wait as long as you want me to Conni, I always appreciate the hard work you put into this project, even well after you were "supposed" to be moving on 🤣

> I am currently working on a workaround for the max query issue here https://github.com/tjdevries/tree-sitter-lua/pull/20
> 
> 
> 
> And it will probably work 

I love the confidence 😁

File: lua/telescope/builtin/init.lua
Comment (Conni2461) [inline lua/telescope/builtin/init.lua:229]: Docs for jumplist are missing

File: lua/telescope/builtin/init.lua
Comment (Conni2461) [inline lua/telescope/builtin/init.lua:85]: There are a lot more "hidden" options which can be found in the entry_maker.
Here for example:
`show_line`, `hide_filename`

**Comment by smithbm2316** [comment]
>  1. `show_line` for tags it removes or add the actual line of the tag. So something like this would make sense: `show_line=true + previewer=false` or `show_line=false` (+ previewer). Because it would be redundant to have both, show the line in results and the line highlighted in the previewer. For `treesitter` its odd. It enables/disables the line:col column (middle one)

Oh perfect that makes sense. I can write something up that will hopefully be better for that.

> 2. Put it in CONTRIBUTING.md :) We have a bunch of stuff to docgen there any. And we just don't merge pickers that don't have any documentation. Althought i have no idea what picker are actually missing. But i also was at that point 4 months ago and now we have like +15 pickers :rofl: 

I can also do that.

> 3. I think `@tag` and `@brief` can only occur once per file

That's what I figured. I've tried something else now for what I was attempting to do. 

> Also you should probably request another review from tami or elianiva. I am not that thrilled to review this PR :rofl: 

Definitely will do that. Still have a couple of things to clean up now before I request that again. Thanks!

**Comment by smithbm2316** [comment]
Think I've finished this next batch of changes and additions. Looks like docgen is working properly now! :tada:

**Comment by joelpalmer** [comment]
This will be a nice update. I just wanted to mention that the wording (_it mostly predates this PR_) around the `grep_string` description is the default behavior. That seems reasonable [here](https://github.com/nvim-telescope/telescope.nvim/pull/783/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R373).

> Searches for the string under your cursor in your current working directory

But, [here](https://github.com/nvim-telescope/telescope.nvim/pull/783/files#diff-79c55de55b65fbaa2b214c1f3d28277a87a5b6ac9e5e0c87168a3341a471e5d4R309-R318) I believe it would be helpful to call out that the string can come from anywhere and be used in any directory using the `search` and `search_dirs` `opts` fields respectively. https://github.com/nvim-telescope/telescope.nvim/blob/25a7ecc289dffaa3d45870b452fa1bfb83253ba9/lua/telescope/builtin/files.lua#L105-L106

For example, I do things like this:

```lua
function M.grep_prompt()
  require('telescope.builtin').grep_string {
    shorten_path = true,
    search = vim.fn.input("Rg> "),
  }
end

function M.grep_notes()
	local opts = {}
	opts.search_dirs = {'~/notes/', '~/.vim/', '~/dotfiles', '~/.config/nvim', '~/vim-dev'}
...
```

Looking forward to the updated docs! This user appreciates your efforts.

File: doc/telescope.txt
Comment (tami5) [inline doc/telescope.txt:288]: 😍

Comment (tami5) [reply doc/telescope.txt:288]: I think we should take this space to talk about what all picker share in common. Like setting custom mappings, picker title ... etc. Not sure if it should be part of this pr as you've already did ALLLOT of amazing work

**Review by tami5 — APPROVED** [review]

File: lua/telescope/builtin/init.lua
Comment (Conni2461) [inline lua/telescope/builtin/init.lua:130]: `git_stash` seems to be missing

File: doc/telescope.txt
Comment (smithbm2316) [reply doc/telescope.txt:288]: I agree we should do that. I think that would probably be a good thing to do in my next PR, which will be focused on cleaning up the README a bit more. Just so we can get this merged ASAP! But I will **definitely** do that, I think this is a great place to put that info

File: lua/telescope/builtin/init.lua
Comment (smithbm2316) [reply lua/telescope/builtin/init.lua:130]: OOPS I forgot that was added recently. Good catch lol

File: doc/telescope.txt
Comment (tami5) [reply doc/telescope.txt:288]: Awesome @smithbm2316 , I'm sooo glad that this is being done. There are a lot of goodies in telescope that I was anxious that will be lost with time. Thank you

File: doc/telescope.txt
Comment (Conni2461) [inline doc/telescope.txt:285]: Maybe we can get here another example here on how it would work with dropdown/themes and additional opts.

See last discussion in gitter: https://gitter.im/nvim-telescope/community

```lua
require'telescope.builtin'.find_files(require('telescope.themes').get_dropdown({
  key = value, 
  ...
})
```

Comment (smithbm2316) [reply doc/telescope.txt:285]: Sure! I can again have that included in the next PR with updates and cleanup to the README.

**Comment by smithbm2316** [comment]
@joelpalmer thank you very much for the suggestion and the kind words! I'd love to add a note about that in the next PR that I'll be making, with updates focused on the README. When I get a chance to get to my desktop and not be responding to these comments from my phone, I'll add that to my list of To Dos in the Documentation board for this repo.

File: README.md
Comment (elianiva) [inline README.md:390]: don't forget to wrap every `<cr>` with backticks, otherwise it will be treated as an invalid HTML and removed from the rendered markdown.

Comment (smithbm2316) [reply README.md:390]: Oh good catch! I'll make sure I revise that before merging.

File: doc/telescope.txt
Comment (smithbm2316) [reply doc/telescope.txt:285]: image

I'm just adding this here because I apparently can't attach screenshots to Github project kanban boards, and want a reference to what we were talking about here that I can link to there to track this

**Comment by smithbm2316** [comment]
Alright, I've made updates to the README and docs to properly wrap any `<keymaps>` with backticks to avoid the markdown not rendering them in the README, as well as added docs for the `git_stash` picker. I think this should be ready to merge now! :tada:

Any last questions or concerns before I squash and merge this?

**Comment by tjdevries** [comment]
I would say looks good overall. Really awesome work!

I think we can probably just ship it and people can submit smaller updates if they have concerns.

**Review by elianiva — APPROVED** [review]

**Pull request merged**
